### PR TITLE
Add unit tests for com.taobao.arthas.boot.DownloadUtils

### PR DESCRIPTION
**Pull request by Braavos6**
I've analysed your codebase and noticed that com.taobao.arthas.boot.DownloadUtils is not fully tested.
I've written some tests for the methods in this class with the help of [Diffblue Cover](https://www.diffblue.com/opensource).

Hopefully, these tests will help you detect any regressions caused by future code changes. If you would find it useful to have additional tests written for this repository, I would be more than happy to look at other classes that you consider important.

### CODE REVIEW

File: boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    public void testgetRepoUrl() {
+        Assert.assertEquals("http", DownloadUtils.getRepoUrl("https/", true));
+        Assert.assertEquals("https://maven.aliyun.com/repository/public", DownloadUtils.getRepoUrl("aliyun", false));
+    }
+
+    @Test
+    public void testreadMavenMetaData() {
+        Assert.assertNull(DownloadUtils.readMavenMetaData("", true));
+
+        Assert.assertNotNull(DownloadUtils.readMavenMetaData("center", false));
```
Comment (peterschrammel) [inline boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java]: More checks on the return value?

File: boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    @Test
+    public void testdownArthasPackaging() throws IOException, SAXException, ParserConfigurationException {
+        DownloadUtils.downArthasPackaging("http://repo1.maven.org/maven2", true, "3.0.4", "savePath");
+    }
+}
```
Comment (peterschrammel) [inline boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java]: missing newline at the end of the file

File: boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    @Test
+    public void testreadMavenMetaData() {
+        Assert.assertNull(DownloadUtils.readMavenMetaData("", true));
+
+        Assert.assertNotNull(DownloadUtils.readMavenMetaData("center", false));
+    }
+
+    @Test
+    public void testdownArthasPackaging() throws IOException, SAXException, ParserConfigurationException {
+        DownloadUtils.downArthasPackaging("http://repo1.maven.org/maven2", true, "3.0.4", "savePath");
```
Comment (peterschrammel) [inline boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java]: What is tested here?

File: boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java (hunk 1 above)
Comment (Braavos6) [reply boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java]: I added a few more asserts for null/not null and equals to empty string. However, I am unsure how to check the return value as when printed it produced this (it came out as code) -

<?xml version="1.0" encoding="UTF-8"?>
<metadata>
  <groupId>com.taobao.arthas</groupId>
  <artifactId>arthas-packaging</artifactId>
  <versioning>
    <latest>3.1.0</latest>
    <release>3.1.0</release>
    <versions>
      <version>3.0.0-RC</version>
      <version>3.0.3</version>
      <version>3.0.4</version>
      <version>3.0.5</version>
      <version>3.1.0</version>
    </versions>
    <lastUpdated>20190213182725</lastUpdated>
  </versioning>
</metadata>

File: boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java (hunk 3 above)
Comment (Braavos6) [reply boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java]: That was more for the code coverage as it provided 39% of it. I wasn't sure if to keep it or not, but i will remove it.

File: boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java (hunk 2 above)
Comment (Braavos6) [reply boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java]: Thanks I added the newline

File: boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    @Test
+    public void testreadAllMavenVersion() {
+        Assert.assertEquals(new ArrayList<String>(), DownloadUtils.readAllMavenVersion(""));
+        Assert.assertEquals(new ArrayList<String>(), DownloadUtils.readAllMavenVersion(null));
+    }
+
+    @Test
+    public void testgetRepoUrl() {
```
Comment (petersear) [inline boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java]: Do the first two tests in this method cover any lines which are not covered by the other three?

File: boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        Assert.assertEquals("", DownloadUtils.getRepoUrl(" ", true));
+        Assert.assertEquals("-)m", DownloadUtils.getRepoUrl("-)m/", true));
+        Assert.assertEquals("http", DownloadUtils.getRepoUrl("https/", true));
+        Assert.assertEquals("https://repo1.maven.org/maven2", DownloadUtils.getRepoUrl("center", false));
+        Assert.assertEquals("https://maven.aliyun.com/repository/public", DownloadUtils.getRepoUrl("aliyun", false));
+    }
+
+    @Test
+    public void testreadMavenMetaData() {
```
Comment (petersear) [inline boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java]: I think the assertEquals() checks can replace the assertNotNull() checks as they are testing the same code but with a more specific result.

File: boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.ArrayList;
+
+public class DownloadUtilsTest {
+
+    @Test
+    public void testreadMavenReleaseVersion() {
```
Comment (petersear) [inline boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java]: As discussed, capitalize the first letter in the method name to separate it from the "test" prefix.

**Review by petersear — CHANGES_REQUESTED** [review]
As discussed, capitalize the first letter of the original method name in the test method name.
Also, the class source file should be in a file structure com/taobao/arthas/boot, not a folder called com.taobao.arthas.boot (this is not helped by IntelliJ which seems to show both in exactly the same way)

File: boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java (hunk 6 above)
Comment (Braavos6) [reply boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java]: first letters of the method names have been capitalised

File: boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java (hunk 5 above)
Comment (Braavos6) [reply boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java]: assertNotNull() has been replaced with assertEquals()

File: boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java (hunk 4 above)
Comment (Braavos6) [reply boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java]: Removed those asserts

File: boot/src/test/java/com/taobao/arthas/boot/DownloadUtilsTest.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    public void testGetRepoUrl() {
+        Assert.assertEquals("http", DownloadUtils.getRepoUrl("https/", true));
+        Assert.assertEquals("https://repo1.maven.org/maven2", DownloadUtils.getRepoUrl("center", false));
+        Assert.assertEquals("https://maven.aliyun.com/repository/public", DownloadUtils.getRepoUrl("aliyun", false));
+    }
+
+    @Test
+    public void testReadMavenMetaData() throws IOException {
+        Assert.assertEquals(IOUtils.toString(new URL("http://repo1.maven.org/maven2/com/taobao/arthas/arthas-packaging/maven-metadata.xml").openStream()), DownloadUtils.readMavenMetaData("center", true));
+        Assert.assertEquals(IOUtils.toString(new URL("http://archiva-maven-storage-prod.oss-cn-beijing.aliyuncs.com/repository/central/com/taobao/arthas/arthas-packaging/maven-metadata-public.xml?Expires=1555349003&OSSAccessKeyId=LTAIfU51SusnnfCC&Signature=L6XWC%2BKA1wa0g%2BrpqvMP7EqrRzs%3D").openStream()), DownloadUtils.readMavenMetaData("aliyun", true));
```
Comment (petersear) [inline boot/src/test/java/com/taobao/arthas/boot/DownloadUtilsTest.java]: The assert on line 33 is ugly (and very long) whereas the assert on line 32 is less so. Can this be improved?

**Review by petersear — COMMENTED** [review]
You should be able to reference the methods directly, rather than prefixed with the class name (eg. DownloadUtils.readMavenReleaseVersion() ), if you import the tested class.

File: boot/src/test/java/com/taobao/arthas/boot/DownloadUtilsTest.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
+package com.taobao.arthas.boot;
+
+import com.taobao.arthas.common.IOUtils;
```
Comment (petersear) [inline boot/src/test/java/com/taobao/arthas/boot/DownloadUtilsTest.java:3]: Add a blank line after line 3 to separate groups of imports.

**Review by petersear — APPROVED** [review]

File: boot/src/test/java/com/taobao/arthas/boot/DownloadUtilsTest.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    @Test
+    public void testGetRepoUrl() {
+        Assert.assertEquals("http", getRepoUrl("https/", true));
+        Assert.assertEquals("https://repo1.maven.org/maven2", getRepoUrl("center", false));
+        Assert.assertEquals("https://maven.aliyun.com/repository/public", getRepoUrl("aliyun", false));
+    }
+
+    @Test
+    public void testReadMavenMetaData() throws IOException {
+        Assert.assertEquals(IOUtils.toString(new URL("http://repo1.maven.org/maven2/com/taobao/arthas/arthas-packaging/maven-metadata.xml").openStream()), readMavenMetaData("center", true));
```
Comment (peterschrammel) [inline boot/src/test/java/com/taobao/arthas/boot/DownloadUtilsTest.java]: Break lines > 80 chars

File: boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java (hunk 3 above)
Comment (peterschrammel) [reply boot/src/test/java/com.taobao.arthas.boot/DownloadUtilsTest.java]: A test must always check what the method under test has done.

**Review by peterschrammel — APPROVED** [review]

File: boot/src/test/java/com/taobao/arthas/boot/DownloadUtilsTest.java (hunk 9 above)
Comment (Braavos6) [reply boot/src/test/java/com/taobao/arthas/boot/DownloadUtilsTest.java]: Hi Peter,
Would you like me to create a string variable for the url to shorten the assert?

Comment (peterschrammel) [reply boot/src/test/java/com/taobao/arthas/boot/DownloadUtilsTest.java]: for example

**Pull request closed without merging**